vm/src/builtins/float.rs

```diff
@@ -539,7 +539,8 @@ impl Comparable for PyFloat {
 impl Hashable for PyFloat {
     #[inline]
     fn hash(zelf: &crate::Py<Self>, _vm: &VirtualMachine) -> PyResult<hash::PyHash> {
-        Ok(hash::hash_float(zelf.to_f64()))
+        Ok(hash::hash_float(zelf.to_f64())
+            .unwrap_or_else(|| hash::hash_pointer(zelf as *const _ as *const libc::c_void)))
```

Comment:
Casting to raw pointer is somewhat unpleasure thing in Rust code.
If possible, we want to avoid it.
I checked all of the callers of _Py_HashPointer in cpython. Most of them were `&PyObject` value in RustPython except for method and contextvar. In this case, In this case, I prefer to have a function getting `&PyObject` instead of doing raw pointer casting here and there.
And, we already have the pointer value of PyObject. `AsObject::get_id` is the function to get it. (I left a comment about this too)

Reply:
> Casting to raw pointer is somewhat unpleasure thing in Rust code. If possible, we want to avoid it. I checked all of the callers of _Py_HashPointer in cpython. Most of them were `&PyObject` value in RustPython except for method and contextvar. In this case, In this case, I prefer to have a function getting `&PyObject` instead of doing raw pointer casting here and there. And, we already have the pointer value of PyObject. `AsObject::get_id` is the function to get it. (I left a comment about this too)

In cpython, `float('nan')`'s object id is not equals to the object's hashing value due to the bit rotating logic.

```
>>> import math
>>> hash(math.nan)
276386259  // 10000011110010101000111010011
>>> id(math.nan)
4422180144 // 100000111100101010001110100110000
>>> 
```

So, I think it would be good to have `hash_pointer` function somewhere for future compatibility.

Since `&PyObject` type is not available in `common/src` crate, how about moving the two functions under the `vm` crate to avoid the type issue?

Reply:
sure, that's because we didn't implement pointer hash compatibility.
In object.rs, hash is implemented like this
```
    fn slot_hash(zelf: &PyObject, _vm: &VirtualMachine) -> PyResult<PyHash> {
        Ok(zelf.get_id() as _)
    }
```
which is hash_pointer in CPython.

Of course we need new hash function for that, but I mean, not taking raw pointer but object id.
We don't need pointer for shift operation. That's an integer operation. In this PR, we need hash_object_id, which can take `usize`, instead of hash_pointer. We may add hash_pointer to take something other than `usize` for methods and contextvar, but they also will never take a c pointer - because that's not a rust type.

Reply:
I appreciate for your detailed explanation :)

```rust
    fn slot_hash(zelf: &PyObject, _vm: &VirtualMachine) -> PyResult<PyHash> {
        Ok(hash_object_id(zelf.get_id() as _))
    }
```

Would it be good to apply this change in this PR?